---
title: "HIV_Data"
author: "Denglinsui"
date: "2021/3/16"
output: html_document
bibliography: mybib.bib
---

# Background
Recently, understanding the genotype-phenotype correlation provides guidance of treatment choice. @Rhee2006 related HIV-1 protease and reverse transcriptase mutations to in vitro susptibility to 16 antiretroviral drugs. Breifly speaking, we are interested in the gene mutations resistance to each drugs. This task is challenging because the sample size is not large enough compared to the complete gene mutations. A possible solution is taking advantage of expert information and considering a small collection. However, we wish to discover more potential intereting mutations and to suggest candidates for the further experiment. 

Our target is not just making more discoveries but controlling the false discoveries. The common criterions of controlling false discoveries are false discovery rate [@Benjamini1995; @Benjamini2001; @Barber2015] and $k$-familywise error rate[@Holm1979; @Janson2016]. We are going to explore the dataset provided by @Rhee2006 and to exploit more HIV-1 drug resistant phenotype related genotype.

# Explore the HIV-1 Drug Resistance Dataset
The dataset is available at [HIV DRUG RESISTANCE DATABASE](http://hivdb.stanford.edu/pages/published_analysis/genophenoPNAS2006). 

## Antiretroviral Drug
There are four major types of antiretroviral drugs: eight protease inhibitors(PIs), seven nucleoside and one nucleotide reverse transicriptase inhibitors(NRTIs), three nonnucleoside reverse transcriptase inhibitors(NNRTIs) amd one fusion inhibitor. The first three drug classes are included in this dataset.

```{r Drug_Class, echo=FALSE}
rm(list = ls())
library(knitr)
library(ggplot2)
drug_class_all <- c("PI","NRTI","NNRTI")
Drug_extract <- function(drug_class){
  base_url = 'http://hivdb.stanford.edu/pages/published_analysis/genophenoPNAS2006'
  gene_url = paste(base_url, 'DATA', paste0(drug_class, '_DATA.txt'), sep='/')
  tsm_url = paste(base_url, 'MUTATIONLISTS', 'NP_TSM', drug_class, sep='/')
  
  gene_df = read.delim(gene_url, na.string = c('NA', ''), stringsAsFactors = FALSE)

  tsm_df = read.delim(tsm_url, header = FALSE, stringsAsFactors = FALSE)
  names(tsm_df) = c('Position', 'Mutations')

  pos_start = which(names(gene_df) == 'P1')
  Drug_name = paste(names(gene_df)[4:(pos_start-1)],collapse = " ")
  return(list(gene_df = gene_df,
              tsm_df = tsm_df,
              Drug_name = Drug_name,
              Drug_class = drug_class))
}

#PI
res_PI <- Drug_extract("PI")
res_NRTI <- Drug_extract("NRTI")
res_NNRTI <- Drug_extract("NNRTI")

Drug_summary <- data.frame(DrugClass = drug_class_all,
                           Drug = c(res_PI$Drug_name,
                                    res_NRTI$Drug_name,
                                    res_NNRTI$Drug_name))
row.names(Drug_summary) <- NULL
knitr::kable(Drug_summary)
```

## Exploratory Data Analysis
### Gene Mutations
### Data Visualization

Now, let us scrutinize the gene mutations. The complete gene sets varies from different drug class and we illustrate the rough distribution of the times of gene mutation for the three drug class. Compared to the gene mutations with low frequencies, we prefer those with high frequencies due to the value of further exploration and the stability of the result.
```{r Gene_Heatmap, echo=FALSE}
heatmap_plot <- function(Drug){
  gene_df <- Drug$gene_df
  grepl_rows <- function(pattern, df) {
    cell_matches = apply(df, c(1,2), function(x) grepl(pattern, x))
    apply(cell_matches, 1, all)
  }

  pos_start = which(names(gene_df) == 'P1')
  pos_cols = seq.int(pos_start, ncol(gene_df))
  valid_rows = grepl_rows('^(\\.|-|[A-Zid]+)$', gene_df[,pos_cols])
  gene_df = gene_df[valid_rows,]


  # Construct preliminary design matrix.
  muts = c(LETTERS, 'i', 'd')
  X = outer(muts, as.matrix(gene_df[,pos_cols]), Vectorize(grepl))
  X = aperm(X, c(2,3,1))
  dimnames(X)[[3]] <- muts
  
  mut.freq <- apply(X,c(2,3),mean)
  
  #Heatmap
  heatmap(mut.freq, Rowv = NA, Colv=NA, main = paste("Times of Gene Mutations for",Drug$Drug_class))
}

heatmap_plot(res_PI)

heatmap_plot(res_NRTI)

heatmap_plot(res_NNRTI)

```

### Preliminary Exploration
Besides the effect of single mutation, we are also interested in the across effect of several mutations. As discussed above, we have to explore the gene combinations with high frequencies. We can apply `Aprior Algorithm` to detect the association rule, especially the support set. The effect of three drug classes need to be analyzed seperately as before.

* To begin with, we have to flatten the matrix of the gene mutations associated with paticular genes.
```{r Flatten, echo=FALSE}
# Flatten a matrix to a vector with names from concatenating row/column names.
flatten_matrix <- function(M, sep='.') {
  x <- c(M)
  names(x) <- c(outer(rownames(M), colnames(M),
                      function(...) paste(..., sep=sep)))
  x
}

Flatten_X <- function(Drug){
  gene_df <- Drug$gene_df
  muts = c(LETTERS, 'i', 'd')
  pos_start = which(names(gene_df) == 'P1')
  
  X = outer(muts, as.matrix(gene_df[,pos_start:ncol(gene_df)]),
            Vectorize(grepl))
  X = aperm(X, c(2,3,1))
  dimnames(X)[[3]] <- muts
  X = t(apply(X, 1, flatten_matrix))
  mode(X) <- 'numeric'
  return(X)
}

res <- res_PI
X <- Flatten_X(res)
gene_df <- res$gene_df
pos_start <- which(names(gene_df) == 'P1')
Y <- gene_df[,4:(pos_start-1)]
tsm_df <- res$tsm_df
```

* Next, we apply Aprior Algorithm to the data mutation separately. Here we only use PI drugs as an example.
```{r Aprior Algorithm}
library(arules)
library(arulesViz)

MyTrans<-as(X[,colSums(X)!=0],"transactions")
MyRules<-apriori(data=MyTrans,
                 parameter=list(support=0.1,
                                confidence = 0,
                                minlen=2,
                                target="rules"))
#summary(MyTrans)
#itemFrequencyPlot(MyTrans,topN=10,col="lightblue")
#test <- itemFrequency(MyTrans)
#Size <- size(MyTrans)
#inspect(MyRules)

MyRules.sorted<-sort(x=MyRules,by="support",decreasing=TRUE)
inspect(MyRules.sorted)

```

* It should be noted here we do not hope to analyze the association rule among the gene but to find highly frequent itemsets.

* We will do this after detect the effect after detecting the analysis for 

## Knockoff and BH procedure

```{r knockoff_and_bhq}
knockoff_and_bhq <- function (X, y, q) {
  # Log-transform the drug resistance measurements.
  y = log(y)
  
  # Remove patients with missing measurements.
  missing = is.na(y)
  y = y[!missing]
  X = X[!missing,]
  
  # Remove predictors that appear less than 3 times.
  X = X[,colSums(X) >= 3]
  
  # Remove duplicate predictors.
  X = X[,colSums(abs(cor(X)-1) < 1e-4) == 1]
  
  # Run the knockoff filter.
  knock.gen = function(x) create.fixed(x, method='equi')
  result = knockoff.filter(X, y, fdr=fdr, knockoffs=knock.gen, statistic=stat.glmnet_lambdasmax)
  knockoff_selected = names(result$selected)
  
  # Run the knockoff filter.
  knock.gen = function(x) create.second_order(x, method='equi')
  result = knockoff.filter(X, y, fdr=fdr, knockoffs=knock.gen, statistic=stat.glmnet_lambdasmax)
  knockoff_selected_2ndO = names(result$selected)
  
  # Run BHq.
  p = ncol(X)
  lm.fit = lm(y ~ X - 1) # no intercept
  p.values = coef(summary(lm.fit))[,4]
  names(p.values) <- substring(names(p.values),2)
  cutoff = max(c(0, which(sort(p.values) <= fdr * (1:p) / p)))
  bhq_selected = names(which(p.values <= fdr * cutoff / p))
  
  # Union Result
  Union = union(knockoff_selected, union(knockoff_selected_2ndO, bhq_selected))
  list(Knockoff = knockoff_selected,Knockoff_2ndO = knockoff_selected_2ndO,
       BHq = bhq_selected, Union = Union)
}
```

```{r run_fun}
library(knockoff)
fdr = 0.20

Y = gene_df[,4:(pos_start-1)]
results = lapply(Y, function(y) knockoff_and_bhq(X, y, fdr))
```


```{r plot_result}
get_position <- function(x)
  sapply(regmatches(x, regexpr('[[:digit:]]+', x)), as.numeric)

comparisons <- lapply(results, function(drug) {
  lapply(drug, function(selected) {
    positions = unique(get_position(selected)) # remove possible duplicates
    discoveries = length(positions)
    false_discoveries = length(setdiff(positions, tsm_df$Position))
    list(true_discoveries = discoveries - false_discoveries,
         false_discoveries = false_discoveries,
         fdp = false_discoveries / max(1, discoveries),
         pow = (discoveries - false_discoveries) / length(tsm_df$Position))
  })
})

pic <- list()
i <- 1
library(reshape)
for (drug in names(comparisons)) {
  plot_data <- do.call(cbind, comparisons[[drug]])
  plot_data_tmp <- data.frame(Knockoff_Fix = 
                           as.numeric(plot_data[c(1,2),'Knockoff']),
                         Knockoff_2ndO =
                           as.numeric(plot_data[c(1,2),'Knockoff_2ndO']),
                         BHq = as.numeric(plot_data[c(1,2),'BHq']),
                         Union = as.numeric(plot_data[c(1,2),'Union']),
                         Criteria = rownames(plot_data)[1:2])
  
  plot_data_dataframe <- melt(plot_data_tmp, id=c("Criteria"))
  
  colnames(plot_data_dataframe) <- c("Criteria","Method","Discovery")
  pic[[i]] <- 
    ggplot(data=plot_data_dataframe,
                     aes(x=Method, y = Discovery, fill=Criteria))+
    geom_bar(stat="identity")+
    scale_color_brewer(palette = "Set1")+
    labs(title = paste('Resistance to', drug))
  i <- i+1
}

```

### Summary of Result
```{r}

```

# Reference